Getty detangle

parkan opened the pull request:
closes #16

yusefnapora committed:
fill in getty translator implementation

yusefnapora committed:
wip - write getty translator output to transactor

- no artist deduplication
- untested, probably has stupid bugs

parkan commented on mediachain/translation/__init__.py:
```diff
@@ -0,0 +1,10 @@
+
+
+def get_translator(translator_id):
+    if translator_id == GettyTranslator.translator_id():
```
I always forget that python doesn't have `switch` 😞

yusefnapora commented:
So for thumbnails, I'm thinking about adding a method to the Translator interface like

```python
def get_media_locations(raw_metadata, local_metadata_filepath=None):
```

that will return a dict like:
```python
{"thumbnail": {
    "local": "file:///path/to/thumbnail/file.jpg", 
    "remote": "http://foo.getty.com/image/13421.jpg"
    }
}
```

If you give it a local file path, the translator can return the location it expects to find the thumbnail, otherwise it'll try to pull the remote location from the json in `raw_metadata`.  

Then it's up to the writer to read the file, upload to the store, etc.  If the `--download_thumbnails` flag is set, the writer can try to request the file over http.

parkan commented:
Hmm yeah I see the problem. I'm not sure the translator should care about the local thumbnail, though.

yusefnapora commented:
Yeah, I'm not too keen on it either.   The other thought I had was that we could define a separate interface for walking through a local dataset, and it could return the path to the raw metadata files and media files.  So you'd have one of those for getty dumps and it would know where to find the local thumbs.

yusefnapora committed:
separate parse and translate steps, add comments

yusefnapora committed:
add get_media_locations method to Translator

parkan commented:
How about mapping remote urls to local paths at download time (after translation)? It's a bit annoying because we don't preserve the directory structure of the remote, but it could work.

Either way, translator shouldn't know about the local file, we swap it in in the stateful writer.

yusefnapora commented:
The tricky bit is when we've already got the image from a previous bulk download (as in the getty dumps we've been using).  In the old ingest code I'm just rewriting the local file path from e.g. `getty/json/images/A/B/C/D/ABCDEFG.json` to `getty/downloads/thumb/A/B/C/D/ABCDEFG.jpg`  since that's how they're laid out in the dump directory structure.

But that's not really a property of the getty dataset per-se, just our dump creation code, and it doesn't really belong in the translator.  But we need to put that filename rewrite function somewhere, and since it's specific to getty dumps it doesn't really belong in the writer either....

Probably instead of a "directory walker" interface, we should have a "dataset iterator" interface, one instance of which could walk a getty dump (and return local thumbnail paths if present).  Then we could eventually implement iterators that hit a remote API, etc.  That's pretty similar to the split we have in the l-space translator design as well.

parkan commented:
Right, so I was thinking that maybe we can adopt a standard pattern for this, but that's probably not something we should fix. 

I'm also wondering about the case where we may have multiple images for a particular work (like different views of a statue) or other more complex cases, going up to the IIIF arbitrary canvas case. Hmm.

autoencoder commented:
Yeah, the dataset iterator interface in `mc_datasets` uses the `group_num` key to identify images belonging to the same semantic group. Still a rough preliminary interface.

yusefnapora commented:
I need to look through `mc_datasets` some more... maybe we can unify the dataset iterator between the two projects.   For now I've got an iterator interface which yields translated metadata and the locations of local and remote media.  Then the writer will load the media files (or download, if the flag is set), put to the datastore, and stash the refs into the `meta.data` dict in the Artefact before sending to the transactor.

There's also a `process_media` step that for now just assumes that media items named `thumbnail` are images and tries to resize them to 150x150 px.

I tried to make it fairly generic as to what kind of media you might be storing, and how the data is loaded, but like @parkan said we should probably think more deeply about how we want to link to and represent associated media for our records.  But for now, this gets us back where we were before the refactor, with a `thumbnail` link in `meta.data`.

One thing that sucks right now is figuring out which `DatasetIterator` to use... there's a stupid hack in `cli.main` that checks if the translator id starts with `Getty` and load the `GettyDumpIterator`.  But I'm thinking we'll want to dispatch on the translator id + the args to the ingest command to find the right iterator.   So e.g. you could provide an api key instead of a dump dir to use a different iterator.

The other piece that's still missing in this branch is artist deduplication, but I haven't decided yet how I want to handle that with this design...

parkan commented:
I think maybe this is the way to think about this:

* The output from translator is a map<k,v> where v can be a *simple value*, a *mediachain object* or an *asset/external object*
* Simple values are treated literally
* Mediachain objects are idempontently inserted/upserted => yields mediachain/IPLD address
* Assets are resolved and processed => yields HTTP/IPFS address
* Arbitrary fields can be objects or assets so they shouldn't be referenced by specially named fields, I'm thinking either two dictionaries of `field name -> thing` form or maybe special values in the graph?
* Translator makes no assumptions about order of inserts (other than "dependencies" being inserted first) so writer can spool and reorder all entities first, if desired
* Not sure how to specify case-by-case treatment for assets (thumbnail resize)

yusefnapora commented:
so maybe something like:

```python
translator_output = {
  "canonical":  {
    "__mediachain_object__": true,
    "type": "artefact",
    "meta": {  
      "data": {
        "thumbnail": {
          "__mediachain_asset__": true,
          "uri": "http://foo.bar/thumb.jpg"
      }
    },
  },
  "chain": [
    {
      "__mediachain_object__": true,
      "type": "artefactCreatedBy",
      "meta": {},
      "entity": {
        "__mediachain_object__": true,
        "type": "entity",
        "meta": { "data": { "name": "Mr. Snrub" } }
   }
  ]
}
```

and the writer will basically do a depth-first traversal of each mediachain object and send anything with `__mediachain_object__` to the transactor, and `__mediachain_asset__` to the datastore, replacing the inline dicts with multihash refs...

It would have to insert the main `canonical` first, then it could process each cell in the `chain` sequentially, setting its `ref` to the ref of the main canonical.

parkan commented:
lgtm 👍

yusefnapora committed:
wip - use denormalized translator output format

yusefnapora committed:
more wip - closer to working submit with new translator output

yusefnapora committed:
use "raw datastore" for thumbs and raw_ref (works with ipfs)

parkan commented:
@yusefnapora let's rebase this on master

yusefnapora committed:
wip - start factoring out writer code from getty translator (broken)

- [ci skip]

yusefnapora commented:
good call

yusefnapora committed:
fix writer bugs, update cli args

- ingestion works, but no thumbnails or artist deduplication yet

yusefnapora committed:
factor out directory walking into DatasetIterator interface, bring back thumbnail support for getty

yusefnapora committed:
working ingestion with thumbnails using new translator output format

yusefnapora committed:
rm media_loader.py (replaced by asset_loader.py)

yusefnapora committed:
Update readme

yusefnapora committed:
pytest setup, basic validation of translator output

yusefnapora committed:
rebase fixups

yusefnapora committed:
use multihash_ref helper in Writer

yusefnapora committed:
run unit tests on travis ci

yusefnapora committed:
[travis] combine lint and test into single command

yusefnapora committed:
specify pytest versions in setup.py

yusefnapora committed:
rm old getty ingest code

yusefnapora committed:
writer: catch AbortionError instead of NetworkError

yusefnapora committed:
return shared RpcDatastore instance from get_db()

yusefnapora committed:
Merge master into yn-getty-detangle

yusefnapora committed:
disable console log for missing thumbnail asset

yusefnapora committed:
don't try to read assets from non-existent files

yusefnapora committed:
keep local cache of refs written to rpc datastore, avoid duplicate puts

yusefnapora committed:
remove unused data_object classes, simplify MultihashReference conversions

- removes Record, Entity, Artefact, etc, since they're not used in the
  new ingestion code.

- defines __str__ for MultihashReference as the base58 string,
  so you can just do str(some_multihash_ref)

- cleans up some of the conversions in datastore.utils

- adds a MultihashReference.for_content class method to get a ref
  for a byte string

vyzo committed:
parallel ingestion script

vyzo committed:
ingest-parallel: wait for children to complete!

vyzo commented:
Added parallel ingestion script

yusefnapora committed:
Only set metaSource and raw_ref for base canonical, not related entities

yusefnapora committed:
catch and log grpc errors during ingestion
